runtime/src/util/map.rs

```diff
@@ -0,0 +1,198 @@
+        self.hamt.flush().with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+            format!("failed to flush HAMT '{}'", self.name)
+        })
+    }
+
+    /// Returns a reference to the value associated with a key, if present.
+    pub fn get(&self, key: &K) -> Result<Option<&V>, ActorError> {
+        let k = key.to_bytes().context_code(ExitCode::USR_ASSERTION_FAILED, "invalid key")?;
+        self.hamt.get(&k).with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+            format!("failed to get from HAMT '{}'", self.name)
+        })
+    }
+
+    /// Inserts a key-value pair into the map.
+    /// Returns any value previously associated with the key.
+    pub fn set(&mut self, key: &K, value: V) -> Result<Option<V>, ActorError>
+    where
+        V: PartialEq,
+    {
+        let k = key.to_bytes().context_code(ExitCode::USR_ASSERTION_FAILED, "invalid key")?;
+        self.hamt.set(k.into(), value).with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+            format!("failed to set in HAMT '{}'", self.name)
+        })
+    }
+
+    /// Inserts a key-value pair only if the key does not already exist.
+    /// Returns whether the map was modified (i.e. key was absent).
+    pub fn set_if_absent(&mut self, key: &K, value: V) -> Result<bool, ActorError>
+    where
+        V: PartialEq,
+    {
+        let k = key.to_bytes().context_code(ExitCode::USR_ASSERTION_FAILED, "invalid key")?;
+        self.hamt
+            .set_if_absent(k.into(), value)
+            .with_context_code(ExitCode::USR_ILLEGAL_STATE, || {
+                format!("failed to set in HAMT '{}'", self.name)
+            })
+    }
+
+    /// Iterates over all key-value pairs in the map.
+    pub fn for_each<F>(&self, mut f: F) -> Result<(), ActorError>
+    where
+        // Note the result type of F uses ActorError.
+        // The implementation will extract and propagate any ActorError
+        // wrapped in a hamt::Error::Dynamic.
+        F: FnMut(K, &V) -> Result<(), ActorError>,
+    {
+        match self.hamt.for_each(|k, v| {
+            let key = K::from_bytes(k).context_code(ExitCode::USR_ILLEGAL_STATE, "invalid key")?;
+            f(key, v).map_err(|e| anyhow!(e))
+        }) {
+            Ok(_) => Ok(()),
+            Err(hamt_err) => match hamt_err {
+                hamt::Error::Dynamic(e) => match e.downcast::<ActorError>() {
+                    Ok(ae) => Err(ae),
+                    Err(e) => Err(ActorError::illegal_state(format!(
+                        "error traversing HAMT {}: {}",
+                        self.name, e
+                    ))),
+                },
+                e => Err(ActorError::illegal_state(format!(
+                    "error traversing HAMT {}: {}",
+                    self.name, e
+                ))),
+            },
+        }
+    }
+}
+
+impl MapKey for u64 {
+    fn from_bytes(b: &[u8]) -> Result<Self, String> {
+        let (v, _) = unsigned_varint::decode::u64(b).map_err(|e| e.to_string())?;
```

Comment:
I would validate that `b` doesn't contain any extra bytes.

Reply:
Done